rs/orchestrator/src/upgrade.rs

```diff
@@ -136,20 +140,70 @@ impl Upgrade {
     pub(crate) async fn check(&mut self) -> OrchestratorResult<Option<SubnetId>> {
         let latest_registry_version = self.registry.get_latest_version();
         // Determine the subnet_id using the local CUP.
-        let (subnet_id, local_cup_proto, local_cup) =
-            if let Some(proto) = self.cup_provider.get_local_cup_proto() {
-                let cup = CatchUpPackage::try_from(&proto).expect("deserializing CUP failed");
-                let subnet_id =
-                    get_subnet_id(&*self.registry.registry_client, &cup).map_err(|err| {
-                        OrchestratorError::UpgradeError(format!(
-                            "Couldn't determine the subnet id: {:?}",
-                            err
-                        ))
-                    })?;
-                (subnet_id, Some(proto), Some(cup))
-            } else {
-                // No local CUP found, check registry
-                match self.registry.get_subnet_id(latest_registry_version) {
+        let (subnet_id, local_cup_proto, local_cup) = {
+            let maybe_proto = self.cup_provider.get_local_cup_proto();
+            let maybe_cup = maybe_proto.as_ref().and_then(|proto| {
+                CatchUpPackage::try_from(proto)
+                    .inspect_err(|err| {
+                        error!(self.logger, "Failed to deserialize CatchUpPackage: {}", err);
+                    })
+                    .ok()
+            });
+
+            match (&maybe_cup, &maybe_proto) {
+                (Some(cup), _) => {
+                    let subnet_id =
+                        get_subnet_id(&*self.registry.registry_client, cup).map_err(|err| {
+                            OrchestratorError::UpgradeError(format!(
+                                "Couldn't determine the subnet id: {:?}",
+                                err
+                            ))
+                        })?;
+                    (subnet_id, maybe_proto, maybe_cup)
+                }
+                (None, Some(proto)) => {
+                    // We found a local CUP proto that we can't deserialize. This may only happen
+                    // if this is the first CUP we are reading on a new replica version after an
+                    // upgrade. This means we have to be an assigned node, otherwise we would have
```

Comment:
I wonder whether this is always true: couldn't there be a bug that leads to the replica being unable to deserialize a CUP even if it was written with the same version?

Reply:
In order to persist a CUP we first need to download (deserialize, verify) it from one of the peers or the own replica [here](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/orchestrator/src/catch_up_package_provider.rs?L178-201). If verification of the signature succeeds, it should mean that 2f+1 nodes were able to correctly serialize and deserialize it on the current version

Reply:
For a registry CUP we also need to do the full roundtrip first before persisting it as the local CUP. We first derive the CUP from registry, then serialize it [here](https://github.com/dfinity/ic/pull/572/files#diff-031a135c33914e6d5de51d414a3effc887e6b2d14e8789730ed721fdf381922eL288) and late deserialize it again [here](https://github.com/dfinity/ic/pull/572/files#diff-031a135c33914e6d5de51d414a3effc887e6b2d14e8789730ed721fdf381922eR298).

Reply:
Thank you. Makes sense.